Pxb 8030 to 8029 pxb 2866 8.0

altmannmarcelo opened the pull request.

altmannmarcelo committed:
Implemented PXB-2866 Make PXB 8.0.30 be able to backup 8.0.29 (and prior) server

https://jira.percona.com/browse/PXB-2866

 - Adjusted redo log logic to populate start and end lsn of old redo log
layout.
 - Changed redo_log.cc to use log_sys->m_files_ctx instead of creating a
new Log_files_context each time. log_sys->m_files_ctx will already have
the correct Log_files_ruleset (CURRENT or PRE_8_0_30) that can be used
to identify if we are dealing with ib_logfile or #ib_redo layout.
 - Adjusted lsn calculation on prepare when checkpoint fall into last
block. PRE 8.0.30 were writting LOG_BLOCK_CHECKPOINT_NO on 8th byte of
block header, while on 8.0.30 we write LOG_BLOCK_EPOCH_NO. In cases
where we have two checkpoints faling in the same file (we dont re write
the header) when we reach to the next checkpoint and it falls into a new
block, LOG_BLOCK_CHECKPOINT_NO(which is stored as
recv_sys->scanned_epoch_no) will jump by two. In this case,
log_block_epoch_no_is_valid will identify the gap and will consider the
block as finished, without parsin it.

 - Adjusted consumer code to stop advancing after log_Status

altmannmarcelo committed:
- Fixed pxb-2710-predict_memory.sh test case

Added a condition to only add new 8.0.30+ variable depending on the
version of backup server.
Also fixed a race condition when backup completes before the last
INSERT has the chance to commit. If the INSERT completes in between the
end of backup and before we run the record_db_state we might see
differences on verify_db_state as the last INSERT will be rolled back
by PXB.

 - Fixed all_files.sh test case

We no longer generate redo files to be carried over on copy-back.
Adjusted test case to not check for ib_logfile when running diff

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/signed)](https://cla-assistant.percona.com:443/percona/percona-xtrabackup?pullRequest=1377) <br/>All committers have signed the CLA.

altmannmarcelo committed:
Fixed encryption of redo logs

Adjusted encryption logic to always read encryption info from
ib_logfile0 in case we are using pre 8.0.30 redo layout.

altmannmarcelo committed:
Offset calculation:

[Xtrabackup] Log block checksum mismatch (block no 0 at lsn 18443264): expected 0, calculated checksum 3965168067 block epoch no: 0

Calculation of offsets have changed for 8.0.30+ as the files now carry
an ever increment ID which is used to determine what offset a specific
LSN belongs to (via file->offset(LSN) function). At pre 8.0.30 we had
to use log_files_real_offset_for_lsn.
Same function will have to be used by PXB, in fact we have it as part
of log0pre_8_0_30.cc. The only gotcha here is that it requires us
to at least know one offset for a given LSN. At previous redo log layout
we kept updating log.current_file_real_offset in order to have that
information. On new layout that information is not used anymore.
However, we still have (on old redo log format) LOG_CHECKPOINT_OFFSET,
which is exposed as m_checkpoint_offset from
log_pre_8_0_30::Checkpoint_header and can be used together with
m_checkpoint_lsn to compound the some file LSN & OFFSET.
As part of the LSN/OFFSET calibration, we require to adjust
find_start_checkpoint_lsn to switch between Log_checkpoint_location and
log_pre_8_0_30::Checkpoint_header objects to extract the required
information.

A new function with same name as recv_find_max_checkpoint was added to
log_pre_8_0_30 namespace to populate Checkpoint_header object.

Fix source_offset adjustment for 8029

the source offset for File read needs to be per file. so the difference is
1. in 8.0.29, source_offset is including the two files. (ie. for all IO purpose, all the files are considered as single file logically)
2. using 8.0.30 compute_real_offset() logic we get a similar offset but for file read, the offset range should be within file_size bytes.

add assertion for redo log catchup and adjust the source_offset for aligned read

use 8029 approach for scan_log_recs

Fix backup and prepare issues with a server after redo log resize

Adjusted IF_XB usage to not throw compilation warning if compiled
without XTRABACKUP

Added code comment for 8.0.30 key changes

altmannmarcelo commented on storage/innobase/include/log0files_io.h:
```diff
@@ -519,6 +528,13 @@ inline uint32_t log_block_get_epoch_no(const byte *log_block) {
   return mach_read_from_4(log_block + LOG_BLOCK_EPOCH_NO);
 }
 
+/** Gets a log block epoch_no. For details: @see LOG_BLOCK_CHECKPOINT_NO.
```
epoch_no in comment and get_checkpoint_no in function name.

altmannmarcelo commented on storage/innobase/log/log0files_finder.cc:
```diff
@@ -396,8 +396,14 @@ Log_files_find_result log_files_find_and_analyze(
                                            LOG_HEADER_FLAG_FILE_FULL),
                 encryption_metadata);
     } else {
+#ifdef XTRABACKUP
+      files.add(file_sizes[i].m_id, file_sizes[i].m_size_in_bytes,
+                file_headers[i].m_header.m_start_lsn, true,
+                encryption_metadata);
+#else
```
this can be removed now that we are not using m_start_lsn for pre 8.0.30

altmannmarcelo commented on storage/innobase/log/log0log.cc:
```diff
@@ -1611,7 +1613,10 @@ static
+  // Before 8.0.30, it is possible that ibdata1 was written lsn < LOG_START_LSN
+  // This can happen on redo log resize
+  ut_a(log_is_data_lsn(flushed_lsn) ||
+       xtrabackup_original_log_format <= Log_format::VERSION_8_0_30);
```
```suggestion
       IF_XB(xtrabackup_original_log_format <= Log_format::VERSION_8_0_30));
```

altmannmarcelo commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -54,36 +57,54 @@ Redo_Log_Reader::Redo_Log_Reader() {
-      Log_file::open(log_files_ctx, checkpoint.m_checkpoint_file_id,
-                     Log_file_access_mode::READ_ONLY,
-                     unused_encryption_metadata, Log_file_type::NORMAL);
+    os_offset_t current_file_real_offset = chkp_header.m_checkpoint_offset;
```
this can be removed and checkpoint_offset_start set to chkp_header.m_checkpoint_offset;

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1138,7 +1138,9 @@ static
     }
 
-    if (!file.contains(checkpoint_in_file.m_checkpoint_lsn)) {
+    if (IF_XB(log.m_files_ctx.m_files_ruleset >
```
IF_XB() should take the && too

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -3842,17 +3844,56 @@ bool meb_scan_log_recs(
-      /* Garbage from a log buffer flush which was made
-      before the most recent database recovery */
+#ifdef XTRABACKUP
+    /**
+     * 8.0.30 records LOG_BLOCK_EPOCH_NO at 8th by of log block header
+     * while pre-8.0.30 we recorded LOG_BLOCK_CHECKPOINT_NO
+     */
+    if (scanned_lsn + data_len > recv_sys->scanned_lsn) {
+      if (xtrabackup_original_log_format >= Log_format::VERSION_8_0_30 &&
+          recv_sys->scanned_epoch_no > 0 &&
+          !log_block_epoch_no_is_valid(block_header.m_epoch_no,
```
Can you write a code comment (am sure the commit message will be lost somehow during squash etc)
If not this place somewhere else
All the changes in 8.0.30 vs 8.0.29 redo logs (not the layout)
1. checkpoint_no vs epoch_no
2. checkpoints falling on same file, header update by epoch nos (or is this same as 1)
3. 8.0.30 doesn't store CHECKPOINT_OFFSET anymore and doesn't use it
4. 8.0.30 doesn't use flush_bit anymore
5. anything else

altmannmarcelo replied:
Done

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -3842,17 +3844,56 @@ bool meb_scan_log_recs(
+      }
     }
+#else
+        if (scanned_lsn + data_len > recv_sys->scanned_lsn &&
+            recv_sys->scanned_epoch_no > 0 &&
+            !log_block_epoch_no_is_valid(block_header.m_epoch_no,
+                                         recv_sys->scanned_epoch_no)) {
```
During prepare, should we also deal with the flush_bit?
Our backup parser uses 80.29 parser which deals with flush_bit. 

Without the flush_bit mask, we read the wrong header_no wrong and we stop early.

satya-bodapati replied:
I mean log_data_block_header_deserialize() should consider FLUSH_BIT_MASK and get proper header_no

altmannmarcelo replied:
Already done.
